In [ ]:
opencv官方教程文档: https://www.cnblogs.com/Undo-self-blog/p/8423851.html

In [8]:
import cv2
import numpy as np 
pic = cv2.imread("./1.aidaishu.jpg")   #--->加载图片的路径必须是英文的，中文的会报错

In [3]:
print(pic.shape)    #(4023,3024,3) 数组形状，像素是4023x3024

(4032, 3024, 3)


In [4]:
print(pic)    # --->说明了图片是个数组，而且是个numpy数组

[[[ 42  77  43]
  [ 42  77  43]
  [ 43  78  44]
  ...
  [208 202 203]
  [209 203 204]
  [209 203 204]]

 [[ 42  77  43]
  [ 42  77  43]
  [ 45  78  44]
  ...
  [208 202 203]
  [209 203 204]
  [209 203 204]]

 [[ 42  77  43]
  [ 44  77  43]
  [ 45  78  44]
  ...
  [208 202 203]
  [209 203 204]
  [209 203 204]]

 ...

 [[133 120 112]
  [133 120 112]
  [133 120 112]
  ...
  [172 177 180]
  [172 177 180]
  [172 175 179]]

 [[130 118 112]
  [130 118 112]
  [130 118 112]
  ...
  [172 177 180]
  [172 175 179]
  [172 175 179]]

 [[130 118 112]
  [130 118 112]
  [130 118 112]
  ...
  [172 175 179]
  [172 175 179]
  [174 175 179]]]


In [9]:
cv2.imshow("chenzhuo",pic) #弹出窗口,里面只支持英文名，中文会乱码。第一个参数是放的弹出窗口的名字，第二个房要显示的图片
cv2.waitKey()              #等待键盘输入，任意输入，触发这个代码
cv2.destroyAllWindows()    #销毁内存

In [11]:
cv2.imshow('pic',pic[::-1,:,:])   #::-1  表示翻转   这三个参数表示的是高度，宽度，颜色(蓝，绿，红)
cv2.waitKey()                     #白色是[255,255,255]，因此颠倒完他还是白的，不会发生变化
cv2.destroyAllWindows()

opencv的尺寸和黑白
-cv2.resize(img,( , ))   --->可以改变图片的尺寸

In [16]:
import cv2

img = cv2.imread('./1.aidaishu.jpg')
img = cv2.resize(img,(378,504))
cv2.imshow('chen',img)
cv2.waitKey(0)   #0--->无限等待;1000毫秒=1秒，如果输入一个数，那么到时间后会自动消失
cv2.destroyAllWindows()

In [12]:
img_gray = cv2.cvtColor(img,code = cv2.COLOR_BGR2GRAY)   #--->转换为灰度图
cv2.imshow('gray',img_gray)
cv2.waitKey(0)
cv2.destroyAllWindows()

HSV模型 
-HSV模型也称六角锥体模型 
-在在HSV模型中，颜色是由色度(Hue)饱和度(Saturation)明度(Value)共同组成的

-色度（Hue）使用角度度量的，范围是从0 ° 0\degree0°到360 ° 360\degree360°（逆时针旋转），比如0 ° / 360 ° 0\degree/360\degree0°/360°代表红色，120 ° 120\degree120°代表绿色，240 ° 240\degree240°代表蓝色。
-饱和度（Saturation）表示颜色接近光谱色的程度。一种颜色，可以看成是某种光谱色与白色混合的结果。其中光谱色所占的比例愈大，颜色接近光谱色的程度就愈高，颜色的饱和度也就愈高（参考百度）。其范围是0到1。
-明度（Value）颜色明亮的程度，对于光源色，明度值与发光体的光亮度有关。其范围是0（暗）到1（明）。
-比RGB更加接近人们对彩色的感知经验，方便进行颜色的对比
-能够较容易地跟踪某种颜色的物体，常用于分割指定颜色的物体
 
RGB的局限性
-RGB颜色空间使用三个颜色分量的线性组合来表示颜色，任何颜色都与这三个分量有关，而且这三个分量是高度相关的，所以连续变换颜色时不够直观，想要对颜色进行调整的时候需要调整这三个分量才行
-自然环境下获取的图像容易受自然光照、遮挡和阴影等情况的影响，即对亮度比较敏感。而RGB颜色空间的三个分量都与亮度密切相关，即只要亮度改变，三个分量都会随之改变，而没有一中更直观的的方式来表现
-人眼对于这三种颜色分量的敏感程度是不一样的，在单色中，人眼对红色最不敏感，蓝色最敏感，故RGB颜色空间是一种均匀性较差的颜色空间。对于某种颜色，我们很难推测出较为精确的三个分量数值来表示
-RGB适合于显示系统，而不适合于图像处理

In [15]:
img_hsv = cv2.cvtColor(img,code = cv2.COLOR_BGR2HSV)     #--->转换为HSV
cv2.imshow('hsv',img_hsv)
cv2.waitKey(0)
cv2.destroyAllWindows()

In [14]:
lower_blue = np.array([110,50,50])   #浅蓝色     这两行定义了HSV颜色空间中蓝色的范围
upper_blue = np.array([130,255,255]) #深蓝色
mask = cv2.inRange(img,lower_blue,upper_blue)    #根据蓝色的范围来标记图片中哪些位置是蓝色也就是lower_blue到upper_blue这一段
print(mask)                                      #inRange是否在这个范围，只要在这个范围中，都属于蓝色
cv2.imshow('blue',mask)                          #mask中的值是逻辑值0和1，值在blue的范围中就为1，否则就是0
cv2.waitKey()
cv2.destroyAllWindows()

[[0 0 0 ... 0 0 0]
 [0 0 0 ... 0 0 0]
 [0 0 0 ... 0 0 0]
 ...
 [0 0 0 ... 0 0 0]
 [0 0 0 ... 0 0 0]
 [0 0 0 ... 0 0 0]]


In [19]:
res = cv2.bitwise_and(img,img,mask=mask)
cv2.imshow('after_1',res)
cv2.waitKey()
cv2.destroyAllWindows()